### PR TITLE
Plotting refactor

### PR DESCRIPTION
**Pull request by kdesoto-astro**
## Change Description
HUGE plotting refactor + additional plotting functions. Added plotting test functions are at about 60-70% coverage right now.

## Code Quality
- [x] I have read the Contribution Guide
- [x] My code follows the code style of this project
- [x] My code builds (or compiles) cleanly without any errors or warnings
- [x] My code contains relevant comments and necessary documentation

### CODE REVIEW

File: src/superphot_plus/plotting/sampling_results.py [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+    feature_means_comb = np.vstack((feature_means, feature_means_filler))
+    labels_comb = np.append(labels_ordered, labels_filler)
+    features_smote_comb, labels_smote_comb = oversample_minority_classes(feature_means_comb, labels_comb)
+
+    features_smote = features_smote_comb[labels_smote_comb == allowed_types[0]]
+    labels_smote = labels_smote_comb[labels_smote_comb == allowed_types[0]]
+
+    params, save_labels = param_labels(
+        [
+            "g",
```
Comment (delucchi-cmu) [inline src/superphot_plus/plotting/sampling_results.py:175]: Should this use an `aux_bands` argument?

File: src/superphot_plus/plotting/sampling_results.py [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+        Prior object to overlay prior distributions. Defaults to ZTF's priors.
+    """
+    allowed_types = SnClass.all_classes()
+
+    goal_per_class = OVERSAMPLE_SIZE
+    features_gaussian, labels_gaussian = oversample_using_posteriors(names, labels, goal_per_class, fits_dir)
+
+    params, _ = param_labels(
+        [
+            "g",
```
Comment (delucchi-cmu) [inline src/superphot_plus/plotting/sampling_results.py:254]: If this *should* be aux bands, can use the priors.aux_bands

File: src/superphot_plus/plotting/sampling_results.py [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+    os.makedirs(os.path.join(save_dir, "combined_2d_posteriors"), exist_ok=True)
+    labels_to_class, classes_to_labels = SnClass.get_type_maps()
+    # pt_colors = ["r", "c", "k", "m", "g"] # keep for TODO
+
+    features, labels = oversample_using_posteriors(names, labels, OVERSAMPLE_SIZE, fits_dir)
+
+    print(names)
+    params, save_labels = param_labels(
+        [
+            "g",
```
Comment (delucchi-cmu) [inline src/superphot_plus/plotting/sampling_results.py:371]: I'm just going to comment on all the lines like this, so I look more productive =]

File: src/superphot_plus/plotting/sampling_results.py [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+        Where to save the output figures.
+    overlay_gaussians : boolean, optional
+        Whether to overlay Gaussian estimate of distribution. Defaults to True.
+    """
+    os.makedirs(os.path.join(save_dir, "posterior_hists"), exist_ok=True)
+    posteriors, labels = oversample_using_posteriors(names, labels, OVERSAMPLE_SIZE, fit_folder)
+
+    params, save_labels = param_labels(
+        [
+            "g",
```
Comment (delucchi-cmu) [inline src/superphot_plus/plotting/sampling_results.py:425]: =]

File: tests/superphot_plus/plotting/test_plot_lightcurves.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        "tau_fall_g",
+        "extra_sigma_g",
+    ]
+    return {
+        param: np.random.rand(3, 20) if batch else np.random.rand(1, 20).flatten() for param in param_list
+    }
+
+
+def test_plot_lc_fit(single_ztf_sn_id, ztf_priors, test_data_dir, tmp_path):
+    """Tests plot_lc_fit() and my extension, plot_sampling_lc_fit()."""
```
Comment (delucchi-cmu) [inline tests/superphot_plus/plotting/test_plot_lightcurves.py:39]:
```suggestion
    """Tests plot_lc_fit() and by extension, plot_sampling_lc_fit()."""
```

File: tests/superphot_plus/plotting/test_plot_sampling_results.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    plot_sampling_trace_numpyro,
+    plot_corner_plot_all,
+    compare_oversampling,
+    plot_combined_posterior_space,
+    plot_param_distributions,
+    plot_oversampling_1d,
+)
+
+
+def generate_dummy_posterior_sample_dict(batch=False):
```
Comment (delucchi-cmu) [inline tests/superphot_plus/plotting/test_plot_sampling_results.py:17]: If you want to be fancy, you could turn this into a pytest fixture by adding it in the conftest file and putting a fixture annotation on top. Then, you can use it with lines like:

```
def test_plot_posterior_hist_numpyro_dict(tmp_path, dummy_posterior_sample_dict):
  ...
  plot_posterior_hist_numpyro_dict(dummy_posterior_sample_dict, parameter, tmp_path)
```

You'd likely want one version that's plain, and one that's batched.

File: tests/superphot_plus/plotting/test_format_params.py
Comment (delucchi-cmu) [inline tests/superphot_plus/plotting/test_format_params.py:1]: =[

Comment (kdesoto-astro) [reply tests/superphot_plus/plotting/test_format_params.py:1]: Oops - I forgot to write these haha

File: src/superphot_plus/plotting/classifier_results.py [hunk 7]
```diff
@@ -0,0 +1,701 @@
...
+    plt.savefig(os.path.join(save_dir, "appm_hist_compare.pdf"))
+    plt.close()
+
+
+def plot_chisquared_vs_accuracy(pred_spec_fn, pred_phot_fn, fits_dir, save_dir):
+    """
+    Plot chi-squared value histograms for both the spectroscopic and photometric
+    datasets, and plot spec chi-squared as a function of classification accuracy.
+
+    TODO: IN PROGRESS
```
[the remark this answers is not in the document]
Comment (kdesoto-astro) [reply src/superphot_plus/plotting/classifier_results.py:613]: Yes! It requires some changes for the new calculate_neg_chi_squareds changes I implemented a while back.

File: src/superphot_plus/plotting/classifier_results.py [hunk 8]
```diff
@@ -0,0 +1,701 @@
...
+    )
+
+    # import phot dataframe
+    names_phot, pred_label_alerce, _, pred_class_phot = read_probs_csv(phot_probs_csv)
+    pred_class_phot_alerce = np.array([labels_to_class[x] for x in pred_label_alerce])
+
+    cm_p = confusion_matrix(true_class_spec, pred_class_spec, normalize="pred")
+    cm_p_alerce = confusion_matrix(true_class_alerce, pred_class_spec_alerce, normalize="pred")
+
+    true_fracs = np.array([len(true_classes[true_class_spec == i]) / len(true_class_spec) for i in range(5)])
```
[the remark this answers is not in the document]
Comment (kdesoto-astro) [reply src/superphot_plus/plotting/classifier_results.py:51]: good idea!

File: tests/superphot_plus/plotting/test_format_params.py
Comment (kdesoto-astro) [reply tests/superphot_plus/plotting/test_format_params.py:1]: Ok added!

File: src/superphot_plus/plotting/classifier_results.py [hunk 9]
```diff
@@ -0,0 +1,701 @@
...
+        "ALeRCE (corr.)",
+    ]
+    width = 0.6
+
+    frac_df = pd.read_csv(saved_cf_file)
+
+    true_fracs, pred_fracs, pred_fracs_corr, alerce_fracs, alerce_fracs_corr = frac_df.to_numpy().T
+
+    # Plot YSE class fractions too
+    yse_counts = np.array([314, 107, 15, 2, 32])
```
[the remark this answers is not in the document]
Comment (kdesoto-astro) [reply src/superphot_plus/plotting/classifier_results.py:112]: Very much so! Moving to utils.py

File: src/superphot_plus/plotting/classifier_results.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kdesoto-astro) [reply src/superphot_plus/plotting/classifier_results.py:613]: Postponing to next PR though

**Review by delucchi-cmu — COMMENTED** [review]
Looks good! I will merge in some changes to fix failing unit tests.

**Review by delucchi-cmu — APPROVED** [review]

**Pull request merged**